### PR TITLE
Migrate MultiPPIMeanEstimator onto the engine pattern

### LINKED ISSUE #422: Extract a `_validate_is_2d` primitive from `_validate_y_proxies`
# Extract a `_validate_is_2d` primitive from `_validate_y_proxies`

## Background

`glide/core/validation.py`'s `_validate_y_proxies` inlines its 2D shape check (`if y_proxies.ndim != 2: raise ValueError(...)`) directly in its own body, so nothing else can reuse that exact check without copy-pasting the condition and its error message. A follow-up ticket (`ref-multi-ppi-engine.md`) needs a new `MultiPPIMeanEngine.preprocess` to run that same 2D check on `y_proxies`, without pulling in `_validate_y_proxies`'s other checks (NaN, per-column non-constant), which must not all apply at the engine level. Goal: extract the shape check into a standalone `_validate_is_2d(array, name)` primitive that `_validate_y_proxies` calls, with zero behavior change for any existing caller.

## Design choices

Follow the shape of the other generic primitives already in `validation.py` (e.g. `_validate_non_constant`, `_validate_bounds`): a small function taking the array plus a `name` string used in the error message, placed alongside those primitives rather than next to `_validate_y_proxies`, since it is general-purpose and not proxy-specific. `_validate_y_proxies` calls `_validate_is_2d(y_proxies, "y_proxies")` as its first statement — a direct call, not a wrapped one-liner delegate.

## Implementation

**Files affected:**

| File | What changes |
|---|---|
| `glide/core/validation.py` | Add `_validate_is_2d(array: NDArray, name: str) -> None`. `_validate_y_proxies` calls it as its first statement instead of inlining the check. |
| `tests/unit/core/test_validation.py` | New `_validate_is_2d` section with its own valid-input and raises tests. `test_validate_y_proxies_1d_raises` is removed (that behavior is now `_validate_is_2d`'s own responsibility, tested there); `test_validate_y_proxies_delegates_to_validation` gains a `mock_validate_is_2d` assertion, following the same pattern already used there for `mock_validate_has_no_nan` and `mock_validate_non_constant`. |

**Before / After:**

```python
# Before
def _validate_y_proxies(
    y_proxies: NDArray, identifier: Optional[Hashable] = None, identifier_kind: str = "stratum"
) -> None:
    if y_proxies.ndim != 2:
        raise ValueError(f"'y_proxies' must be a 2D array; got shape {y_proxies.shape!r}.")
    _validate_has_no_nan(y_proxies, "y_proxies")
    identifier_part = f" in {identifier_kind} '{identifier}'" if identifier is not None else ""
    for m in range(y_proxies.shape[1]):
        _validate_non_constant(
            y_proxies[:, m],
            f"'y_proxies' column {m} values are constant{identifier_part}.",
        )


# After
def _validate_is_2d(array: NDArray, name: str) -> None:
    if array.ndim != 2:
        raise ValueError(f"'{name}' must be a 2D array; got shape {array.shape!r}.")


def _validate_y_proxies(
    y_proxies: NDArray, identifier: Optional[Hashable] = None, identifier_kind: str = "stratum"
) -> None:
    _validate_is_2d(y_proxies, "y_proxies")
    _validate_has_no_nan(y_proxies, "y_proxies")
    identifier_part = f" in {identifier_kind} '{identifier}'" if identifier is not None else ""
    for m in range(y_proxies.shape[1]):
        _validate_non_constant(
            y_proxies[:, m],
            f"'y_proxies' column {m} values are constant{identifier_part}.",
        )
```

## Corner cases

- The error message text must stay byte-for-byte identical (`"'y_proxies' must be a 2D array; got shape ...".`), since it moves callers (a raises test, plus whatever downstream code matches on it) without any of them noticing the extraction.
- This ticket does not touch `glide/estimators/multi_ppi.py`, `glide/estimators/multi_ppi_core.py`, or anything under `glide/engines/`. `ref-multi-ppi-engine.md` depends on this ticket merging first, since its `MultiPPIMeanEngine.preprocess` calls `_validate_is_2d` directly.
- `_validate_y_proxies`'s remaining callers (`multi_ptd.py`, `stratified_ppi.py`, `clustered_ppi.py`) are unaffected and out of scope; do not change their behavior.

## Acceptance criteria

- [ ] `_validate_is_2d(array, name)` exists in `glide/core/validation.py` and raises `ValueError(f"'{name}' must be a 2D array; got shape {array.shape!r}.")` for a non-2D array.
- [ ] `_validate_y_proxies` calls `_validate_is_2d` instead of inlining the shape check; its externally observable behavior (error messages, existing callers) is unchanged.
- [ ] `make coverage` stays at 100%: both the valid and raising branches of `_validate_is_2d` are exercised directly by `tests/unit/core/test_validation.py`, not only indirectly through `_validate_y_proxies`.

### LINKED ISSUE #423: Extract `MultiPPIMeanEngine` and migrate `MultiPPIMeanEstimator` onto it
# Extract `MultiPPIMeanEngine` and migrate `MultiPPIMeanEstimator` onto it

## Background

PR #397 introduced `glide/engines/`: a per-family engine (`preprocess`, `fit_tuning_parameter`, `compute_mean_and_std`) implementing the `MeanEstimationEngine` protocol (`glide/engines/base.py`), shared between an estimator (one unit = the whole dataset) and a monitor (one unit = one batch). It migrated PPI and Classical only. `glide/estimators/multi_ppi.py` and `glide/estimators/multi_ppi_core.py` still have the pre-#397 shape: the estimator owns a `_preprocess` method that calls validation directly, and `estimate()` calls the `multi_ppi_core` functions itself before round-tripping through `ClassicalMeanEstimator().estimate(...).confidence_interval` to get the classical variance. Goal: migrate Multi-PPI onto the engine pattern, mirroring `PPIMeanEngine`/`PPIMeanEstimator` exactly.

## Design choices

Mirror `PPIMeanEngine` exactly: rename `glide/estimators/multi_ppi_core.py` to `glide/engines/multi_ppi_core.py`, and add a `MultiPPIMeanEngine` in a new `glide/engines/multi_ppi.py`. Rename `_compute_tuning_parameters` (plural) to `_compute_tuning_parameter` (singular) to match `ppi_core.py`'s naming, since a tuning parameter is a single object even though it happens to be vector-valued here.

Two validation checks move for the same reason they moved in `PPIMeanEngine`. Proxy non-constancy is only a precondition of `_compute_tuning_parameter`'s division by proxy variance, so the per-column check moves out of `preprocess` and into `_compute_tuning_parameter`, gated on `power_tuning`. This isn't optional: `glide/monitors/base.py`'s `_detect` loop calls `preprocess` on every batch and always fits batch 0 with `power_tuning=False` (there is no prefix yet), so a check living in `preprocess` would spuriously reject a small batch whose fit never touches proxy variance in the first place. The `y_true` non-constant check isn't needed by a per-batch caller either, so it moves out of the engine entirely and is called directly in `MultiPPIMeanEstimator.estimate()`, exactly where `PPIMeanEstimator.estimate()` now calls it.

## Implementation

**Files affected:**

| File | What changes |
|---|---|
| `glide/estimators/multi_ppi_core.py` | Moved to `glide/engines/multi_ppi_core.py`. `_compute_tuning_parameters` renamed to `_compute_tuning_parameter` and gains the proxy non-constant guard under `power_tuning`. `_compute_mean_estimate`, `_compute_std_estimate` unchanged. |
| `glide/engines/multi_ppi.py` | New. `MultiPPIDataset = Tuple[NDArray, NDArray, NDArray]` and `MultiPPIMeanEngine`, mirroring `glide/engines/ppi.py`. |
| `glide/estimators/multi_ppi.py` | `_preprocess` method removed. `__init__` sets `self._engine = MultiPPIMeanEngine()` and `self._classical_engine = ClassicalMeanEngine()`. `estimate()` delegates to both engines, mirroring `glide/estimators/ppi.py`. Docstring `Parameters`/`Raises` updated to match the relaxed proxy-variance requirement. |
| `tests/unit/estimators/test_multi_ppi_core.py` | Moved to `tests/unit/engines/test_multi_ppi_core.py`, following `tests/unit/engines/test_ppi_core.py`'s structure. Add a case for the new non-constant-column guard under `power_tuning=True`, mirroring `test_compute_tuning_parameter_delegates_to_validation` in `test_ppi_core.py`. |
| `tests/unit/engines/test_multi_ppi.py` | New. `preprocess`/`fit_tuning_parameter`/`compute_mean_and_std` tests, mirroring `tests/unit/engines/test_ppi.py`. |
| `tests/unit/estimators/test_multi_ppi.py` | `test_preprocess_delegates` and `test_preprocess_valid_output` replaced by a single `test_estimate_delegates`, mirroring `tests/unit/estimators/test_ppi.py`, plus a new `test_init_sets_engines`. Other `estimate()` tests are unchanged. |

**Before / After — `glide/estimators/multi_ppi.py`:**

```python
# Before
class MultiPPIMeanEstimator:
    def _preprocess(self, y_true_all: NDArray, y_proxies_all: NDArray) -> Tuple[NDArray, NDArray, NDArray]:
        _validate_equal_lengths(y_true_all, y_proxies_all, names=["y_true", "y_proxies"])
        _validate_y_proxies(y_proxies_all)
        _validate_y_true(y_true_all)
        y_true, y_proxies_labeled, y_proxies_unlabeled, labeled_mask = _split_labeled_unlabeled(
            y_true_all, y_proxies_all
        )
        _validate_sample_sizes(labeled_mask)
        return y_true, y_proxies_labeled, y_proxies_unlabeled

    def estimate(self, y_true, y_proxies, metric_name="Metric", confidence_level=0.95, power_tuning=True):
        y_true_filtered, y_proxies_labeled, y_proxies_unlabeled = self._preprocess(y_true, y_proxies)
        n_labeled = len(y_true_filtered)
        n_unlabeled = len(y_proxies_unlabeled)
        lambdas_ = _compute_tuning_parameters(y_true_filtered, y_proxies_labeled, y_proxies_unlabeled, power_tuning)
        mean = _compute_mean_estimate(y_true_filtered, y_proxies_labeled, y_proxies_unlabeled, lambdas_)
        std = _compute_std_estimate(y_true_filtered, y_proxies_labeled, y_proxies_unlabeled, lambdas_)
        confidence_interval = CLTConfidenceInterval(mean=mean, std=std, confidence_level=confidence_level)
        classical_confidence_interval = ClassicalMeanEstimator().estimate(y_true_filtered).confidence_interval
        effective_sample_size = floor(n_labeled * classical_confidence_interval.var / confidence_interval.var)
        ...  # unchanged: build and return PredictionPoweredMeanInferenceResult


# After
class MultiPPIMeanEstimator:
    def __init__(self) -> None:
        self._engine = MultiPPIMeanEngine()
        self._classical_engine = ClassicalMeanEngine()

    def estimate(self, y_true, y_proxies, metric_name="Metric", confidence_level=0.95, power_tuning=True):
        multi_ppi_dataset = self._engine.preprocess(y_true, y_proxies)
        y_true_labeled, _, y_proxies_unlabeled = multi_ppi_dataset
        _validate_non_constant(y_true_labeled, "'y_true' labeled values are constant.")

        tuning_parameter = self._engine.fit_tuning_parameter(multi_ppi_dataset, power_tuning)
        mean, std = self._engine.compute_mean_and_std(multi_ppi_dataset, tuning_parameter)
        confidence_interval = CLTConfidenceInterval(mean=mean, std=std, confidence_level=confidence_level)

        _, classical_std = self._classical_engine.compute_mean_and_std(y_true_labeled, None)
        n_labeled, n_unlabeled = len(y_true_labeled), len(y_proxies_unlabeled)
        effective_sample_size = floor(n_labeled * classical_std**2 / std**2)
        ...  # unchanged: build and return PredictionPoweredMeanInferenceResult
```

**New — `glide/engines/multi_ppi.py`, mirroring `glide/engines/ppi.py`:**

```python
MultiPPIDataset = Tuple[NDArray, NDArray, NDArray]


class MultiPPIMeanEngine:
    def preprocess(self, y_true: NDArray, y_proxies: NDArray) -> MultiPPIDataset:
        _validate_equal_lengths(y_true, y_proxies, names=["y_true", "y_proxies"])
        _validate_is_2d(y_proxies, "y_proxies")
        _validate_has_no_nan(y_proxies, "y_proxies")
        y_true_labeled, y_proxies_labeled, y_proxies_unlabeled, labeled_mask = _split_labeled_unlabeled(
            y_true, y_proxies
        )
        _validate_sample_sizes(labeled_mask)
        return y_true_labeled, y_proxies_labeled, y_proxies_unlabeled

    def fit_tuning_parameter(self, dataset: MultiPPIDataset, power_tuning: bool) -> NDArray:
        y_true, y_proxies_labeled, y_proxies_unlabeled = dataset
        tuning_parameter = _compute_tuning_parameter(y_true, y_proxies_labeled, y_proxies_unlabeled, power_tuning)
        return tuning_parameter

    def compute_mean_and_std(self, dataset: MultiPPIDataset, tuning_parameter: NDArray) -> Tuple[float, float]:
        y_true, y_proxies_labeled, y_proxies_unlabeled = dataset
        mean = _compute_mean_estimate(y_true, y_proxies_labeled, y_proxies_unlabeled, tuning_parameter)
        std = _compute_std_estimate(y_true, y_proxies_labeled, y_proxies_unlabeled, tuning_parameter)
        return mean, std
```

**Before / After — `glide/engines/multi_ppi_core.py` (renamed to singular, gains the guard):**

```python
# Before — glide/estimators/multi_ppi_core.py
def _compute_tuning_parameters(y_true, y_proxies_labeled, y_proxies_unlabeled, power_tuning) -> NDArray:
    n_proxies = y_proxies_labeled.shape[1]
    if not power_tuning:
        lambdas_ = np.full(n_proxies, 1 / np.sqrt(n_proxies))
        return lambdas_
    n_labeled = len(y_true)
    n_unlabeled = len(y_proxies_unlabeled)
    y_proxies_all = np.vstack([y_proxies_labeled, y_proxies_unlabeled])
    proxy_cov_matrix = np.atleast_2d(np.cov(y_proxies_all, rowvar=False, ddof=1))
    ...  # unchanged from here


# After — glide/engines/multi_ppi_core.py
def _compute_tuning_parameter(y_true, y_proxies_labeled, y_proxies_unlabeled, power_tuning) -> NDArray:
    n_proxies = y_proxies_labeled.shape[1]
    if not power_tuning:
        lambdas_ = np.full(n_proxies, 1 / np.sqrt(n_proxies))
        return lambdas_
    n_labeled = len(y_true)
    n_unlabeled = len(y_proxies_unlabeled)
    y_proxies_all = np.vstack([y_proxies_labeled, y_proxies_unlabeled])
    for m in range(n_proxies):
        _validate_non_constant(y_proxies_all[:, m], f"'y_proxies' column {m} values are constant.")
    proxy_cov_matrix = np.atleast_2d(np.cov(y_proxies_all, rowvar=False, ddof=1))
    ...  # unchanged from here
```

## Corner cases

- `power_tuning=False` now accepts a constant proxy column, since the non-constant check no longer runs during `preprocess`. This mirrors the relaxation PR #397 already made for `PPIMeanEstimator`: update the `y_proxies` parameter docstring to "Each column must have nonzero variance when `power_tuning=True`." and the `Raises` entry to "If any column of `y_proxies` is constant (with `power_tuning=True`)."
- Grep confirms only `glide/estimators/multi_ppi.py` and `tests/unit/estimators/test_multi_ppi_core.py` import `glide.estimators.multi_ppi_core` today; no other module needs updating when it moves.
- `_validate_y_proxies` (the function extracted in `ref-validate-is-2d.md`) is unaffected by this ticket and keeps raising for a constant column unconditionally for its one remaining caller, `multi_ptd.py`. This ticket only changes what `MultiPPIMeanEngine.preprocess` calls — never touch `_validate_y_proxies` itself here.
- The "`y_true` contains only NaN" case is no longer a distinct check (as with PPI): it is now caught by `_validate_sample_sizes` raising "Too few labeled or unlabeled samples", since an all-NaN `y_true` leaves 0 labeled samples. Drop the standalone docstring line for it, mirroring `PPIMeanEstimator`'s current `Raises` list.
- `feat-asymptotic-multi-pprm.md` depends on `MultiPPIMeanEngine` existing; do not start it before this ticket merges.

## Acceptance criteria

- [ ] `MultiPPIMeanEngine` exists in `glide/engines/multi_ppi.py` and implements `MeanEstimationEngine[MultiPPIDataset, NDArray]`.
- [ ] `glide/estimators/multi_ppi_core.py` no longer exists; its functions live at `glide/engines/multi_ppi_core.py`.
- [ ] `MultiPPIMeanEstimator` has no `_preprocess` method; `estimate()` delegates to `self._engine` and `self._classical_engine`.
- [ ] `MultiPPIMeanEstimator().estimate(y_true, y_proxies, power_tuning=False)` succeeds when a column of `y_proxies` is constant.
- [ ] `multi_ptd.py`'s use of `_validate_y_proxies` still rejects a constant proxy column regardless of any `power_tuning`-like flag (unaffected by this ticket).

### PR DESCRIPTION
**Pull request by pravallikamavilla**
### Description

- What does this PR do?
Extracts `_validate_is_2d` as a standalone validation helper and reuses it in `_validate_y_proxies`. Then migrates `MultiPPIMeanEstimator` onto the engine pattern introduced in PR #397: `glide/estimators/multi_ppi_core.py` moves to `glide/engines/multi_ppi_core.py` (with `_compute_tuning_parameters` renamed to `_compute_tuning_parameter`), and a new `MultiPPIMeanEngine` (`glide/engines/multi_ppi.py`) mirrors `PPIMeanEngine`. `MultiPPIMeanEstimator` no longer owns a `_preprocess` method; `estimate()` now delegates to `self._engine` and `self._classical_engine`. As with `PPIMeanEstimator`, proxy non-constancy is now checked lazily inside `_compute_tuning_parameter` (only when `power_tuning=True`) rather than eagerly during preprocessing.
- Which issue does it close? (use `Closes #<number>`)
Handles these tickets: [ticket 1](https://github.com/orgs/EmertonData/projects/26/views/2?pane=issue&itemId=242230104), [ticket 2](https://github.com/orgs/EmertonData/projects/26/views/2?pane=issue&itemId=242243476). Closes #422 Closes #423 
- Any noteworthy implementation decisions or trade-offs?
We chose to implement both tickets in the same PR, since the second ticket (the `MultiPPIMeanEngine` migration) directly depends on the `_validate_is_2d` extraction from the first.

### Checklist

Quality gates that must be satisfied before requesting a review:

- [x] I have read `CONTRIBUTING.md`
- [x] `make lint` passes
- [x] `make type-check` passes
- [x] `make tests` passes
- [x] `make coverage` reports 100% coverage
- [ ] `make test-notebooks` passes
- [x] New public API has numpy-style docstrings
- [ ] New public API is inserted in the API reference section of the documentation
- [ ] Docs build without warnings (`make doc`)
- [ ] `CHANGELOG.md` updated if the change is user-facing

### LLM usage

Disclose if an LLM was used in writing this PR:
- [ ] No LLM used
- [x] I used an LLM and I went through and validated all the code myself

### CODE REVIEW

File: glide/estimators/multi_ppi.py [hunk 1]
```diff
@@ -115,24 +95,29 @@ def estimate(
...
+        mean, std = self._engine.compute_mean_and_std(multi_ppi_dataset, tuning_parameter)
         confidence_interval = CLTConfidenceInterval(
             mean=mean,
             std=std,
             confidence_level=confidence_level,
         )
-        classical_confidence_interval = ClassicalMeanEstimator().estimate(y_true_filtered).confidence_interval
-        effective_sample_size = floor(n_labeled * classical_confidence_interval.var / confidence_interval.var)
+
+        _, classical_std = self._classical_engine.compute_mean_and_std(y_true_labeled, None)
```
Comment (gmartinon-ed) [inline glide/estimators/multi_ppi.py:117]: Petite différence ici avec ppi.py. Il faut homogénéiser:
- Soit on écrit `_, classical_std = self._classical_engine.compute_mean_and_std(y_true_labeled, None)`
- Soit on écrit `_, classical_std = self._classical_engine.compute_mean_and_std(multi_ppi_dataset[0], None)`

Lequel vous semble le plus approprié et généralisable aux futures refacto similaires ?

File: glide/engines/multi_ppi.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
+from typing import Tuple
+
+from numpy.typing import NDArray
+
+from glide.core.utils import _split_labeled_unlabeled
+from glide.core.validation import _validate_equal_lengths, _validate_has_no_nan, _validate_is_2d, _validate_sample_sizes
+from glide.engines.multi_ppi_core import _compute_mean_estimate, _compute_std_estimate, _compute_tuning_parameter
+
+MultiPPIDataset = Tuple[NDArray, NDArray, NDArray]
```
Comment (gmartinon-ed) [inline glide/engines/multi_ppi.py:9]: Du coup ici c'est exactement un PPIDataset.
Question que je me pose: est-ce qu'il ne faudrait pas par hasard définir tous nos types datasets dans un fichier python à part, par exemple `engines/dataset_types.py` et ensuite importer ces datasets depuis les modules `engines/*_ppi.py` ? Comme ça, on pourrait réutiliser le `PPIDataset` déjà existant.

Preneur d'une réflexion (assistée par Claude) sur ce sujet :-)

File: glide/engines/multi_ppi.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+class MultiPPIMeanEngine:
+    def preprocess(self, y_true: NDArray, y_proxies: NDArray) -> MultiPPIDataset:
+        _validate_is_2d(y_proxies, "y_proxies")
+        _validate_equal_lengths(y_true, y_proxies, names=["y_true", "y_proxies"])
+        _validate_has_no_nan(y_proxies, "y_proxies")
+        y_true_labeled, y_proxies_labeled, y_proxies_unlabeled, labeled_mask = _split_labeled_unlabeled(
+            y_true, y_proxies
+        )
+        _validate_sample_sizes(labeled_mask)
```
Comment (gmartinon-ed) [inline glide/engines/multi_ppi.py:20]: Question à ce stade: par rapport à ce qu'il y a sur main, on n'utilise ni `_validate_y_true` ni `_validate_y_proxies`. Est-ce normal ? Pourquoi ? Ces fonctions de validation sont-elles amenées à disparaître in fine ?

File: tests/unit/engines/test_ppi.py [hunk 4]
```diff
@@ -22,6 +22,11 @@ def engine():
     return PPIMeanEngine()
 
 
+@pytest.fixture
```
Comment (gmartinon-ed) [inline tests/unit/engines/test_ppi.py:25]: J'allais justement faire la remarque sur une fixture manquante, vous lisez dans mes pensées !

**Review by gmartinon-ed — CHANGES_REQUESTED** [review]
Belle PR!

File: glide/engines/multi_ppi.py (hunk 2 above)
Comment (pravallikamavilla) [reply glide/engines/multi_ppi.py:9]: For now I don't think it's worth it, as it only affects 2 aliases (PPIDataset and MultiPPIDataset). I'd keep the idea in mind for when a third case shows up.

File: glide/engines/multi_ppi.py (hunk 3 above)
Comment (pravallikamavilla) [reply glide/engines/multi_ppi.py:20]: Yes, this is on purpose. Both functions bundle checks that needed to be split apart rather than called at a time.

`_validate_y_proxies` bundles a shape check, a NaN check, and a per-column non-constant check. The non-constant part only makes sense when `power_tuning=True`. So that check now lives inside `_compute_tuning_parameter`, gated by `power_tuning`, mirroring what `ppi_core.py` already does.

`_validate_y_true` bundles an "any labels at all" check and a "labeled values non-constant" check. The non-constant part only protects `effective_sample_size`, which is estimator-only (no monitor computes it), so it's in `estimate()` instead, same as `PPIMeanEstimator`.

For now we're still using both functions elsewhere. As more families migrate to the engine pattern, if we reach a point where they're no longer used anywhere, we'll remove them.

File: glide/estimators/multi_ppi.py (hunk 1 above)
Comment (pravallikamavilla) [reply glide/estimators/multi_ppi.py:117]: Went with the named variable and updated ppi.py to match. Positional indexing into the tuple is more fragile if the field order ever changes.

**Review by gmartinon-ed — APPROVED** [review]
